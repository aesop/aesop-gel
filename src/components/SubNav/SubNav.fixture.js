import { HYPERLINK_STYLE_TYPES } from '~/constants';

const { EXTERNAL_NO_ICON_TEXT_LINK } = HYPERLINK_STYLE_TYPES;

export default {
  links: [
    {
      id: 'about',
      style: EXTERNAL_NO_ICON_TEXT_LINK,
      children: 'About',
      url: 'https://aesop.com.com/au/r/about',
    },
    {
      id: 'philosophy',
      style: EXTERNAL_NO_ICON_TEXT_LINK,
      children: 'Philosophy',
      url: 'https://aesop.com.com/au/r/philosophy-to-products',
    },
  ],
  heading: 'SubNav Title',
};
