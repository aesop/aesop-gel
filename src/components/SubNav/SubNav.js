import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import useWindowHasResized from '~/customHooks/useWindowHasResized';
import { ascertainIsSmallOrMediumOnlyViewport } from '~/utils/viewports';
import Heading from '~/components/Heading';
import Hyperlink from '~/components/Hyperlink';
import Select from '~/components/Select';
import List from '~/components/List';
import styles from './SubNav.module.css';

const SubNav = ({
  className,
  heading,
  headingClassName,
  id,
  isSelect,
  links,
  theme,
}) => {
  useWindowHasResized();

  const currentTheme = useThemeContext(theme, 'dark');
  const classSet = cx(
    styles.base,
    styles[currentTheme],
    { [styles.select]: isSelect },
    className,
  );
  const isSmallOrMediumViewport = ascertainIsSmallOrMediumOnlyViewport();
  const onChange = event => {
    window.location.href = event.target.value;
  };

  return (
    <nav className={classSet}>
      {heading && (
        <Heading
          className={cx(styles.heading, headingClassName)}
          level="3"
          noMargin={true}
          size="small"
          theme={currentTheme}
        >
          {heading}
        </Heading>
      )}
      {isSelect && isSmallOrMediumViewport ? (
        <Select
          isBlock={isSelect}
          name={id}
          onChange={onChange}
          options={links.map(({ children, id, url }) => ({
            id,
            label: children,
            value: url,
          }))}
          theme={currentTheme}
        />
      ) : (
        <List
          items={getLinkItems(links, currentTheme)}
          listItemClassName={styles.item}
          theme={currentTheme}
        />
      )}
    </nav>
  );
};

export function getLinkItems(links, theme) {
  return links.map(({ children, id, hasTargetInNewWindow, style, url }) => ({
    content: (
      <Hyperlink
        className={styles.link}
        hasTargetInNewWindow={hasTargetInNewWindow}
        style={style}
        theme={theme}
        url={url}
      >
        {children}
      </Hyperlink>
    ),
    id,
  }));
}

SubNav.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string,
  isSelect: PropTypes.bool,
  links: PropTypes.arrayOf(PropTypes.object)
    .isRequired /** @TODO hyperlink type */,
  heading: PropTypes.string,
  headingClassName: PropTypes.string,
  theme: PropTypes.oneOf(['dark', 'light']),
};

SubNav.defaultProps = {
  className: undefined,
  id: undefined,
  isSelect: false,
  links: undefined,
  heading: undefined,
  headingClassName: undefined,
  theme: undefined,
};

export default SubNav;
