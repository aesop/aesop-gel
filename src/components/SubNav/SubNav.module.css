@import 'styles/index';

.base {
  margin: 20px 0;
}

.heading {
  font-family: var(--font-family-sans-serif-regular);
  line-height: 1.6;
  transition: all 300ms cubic-bezier(0.215, 0.61, 0.355, 1);
}

.item {
  margin-bottom: 20px;
}

.link {
  font-family: var(--font-family-sans-serif-medium);
  font-size: rem(16px);
  line-height: 1.8;
  transition: color 200ms;

  &:hover {
    text-decoration: underline;
  }

  .dark & {
    color: var(--color-dark-link);
  }

  .light & {
    color: var(--color-light-link);
  }

  .select & {
    font-size: rem(14px);

    &:hover {
      text-decoration: none;
    }
  }

  .select.dark & {
    color: var(--color-dark-label);
  }

  .select.light & {
    color: var(--color-light-label);
  }

  .select.dark &:hover {
    color: var(--color-dark-link);
  }

  .select.light &:hover {
    color: var(--color-light-link);
  }
}
