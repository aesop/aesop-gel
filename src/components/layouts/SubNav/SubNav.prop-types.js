import PropTypes from 'prop-types';

export const HYPERLINK_PROP_TYPE_SHAPE = PropTypes.object;

export const PROP_TYPES = {
  className: PropTypes.string,
  links: PropTypes.arrayOf(HYPERLINK_PROP_TYPE_SHAPE).isRequired, // @TODO hyperlink type
  title: PropTypes.string,
  titleClassName: PropTypes.string,
};

export default PROP_TYPES;
