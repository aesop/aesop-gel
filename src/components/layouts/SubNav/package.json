{
  "name": "SubNav",
  "version": "0.0.0",
  "main": "SubNav.js"
}
