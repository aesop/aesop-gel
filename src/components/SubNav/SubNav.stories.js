import React from 'react';
import { storiesOf } from '@storybook/react';
import { select } from '@storybook/addon-knobs';
import SubNav from './SubNav';
import SubNavFixture from './SubNav.fixture';

storiesOf('Components/SubNav', module).add('Base component', () => (
  <SubNav
    heading={SubNavFixture.heading}
    links={SubNavFixture.links}
    theme={select('theme', ['dark', 'light'], 'dark')}
  />
));
