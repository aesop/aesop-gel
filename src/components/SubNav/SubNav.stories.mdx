import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import SubNav from './SubNav';
import SubNavFixture from './SubNav.fixture';

<Meta title="Components/SubNav" component={SubNav} />

# SubNav

<Preview>
  <Story name="Base component">
    <SubNav
      heading={SubNavFixture.heading}
      id={SubNavFixture.id}
      isHero={knobs.boolean('isHero', false, 'Presentation')}
      isSelect={knobs.boolean('isSelect', false, 'Presentation')}
      links={SubNavFixture.links}
      theme={knobs.select('theme', ['dark', 'light'], 'dark', 'Presentation')}
    />
  </Story>
</Preview>

## API

<Props of={SubNav} />

## Navigation block on Mobile and Tablet

<Preview>
  <Story name="Navigation block">
    <SubNav id="Navigation block" isSelect={true} links={SubNavFixture.links} />
  </Story>
</Preview>
