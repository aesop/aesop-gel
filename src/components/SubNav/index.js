/* istanbul ignore file */
import { SubNav } from './SubNav';

export { SubNav };
