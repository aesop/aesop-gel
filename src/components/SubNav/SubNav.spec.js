import React from 'react';
import renderer from 'react-test-renderer';
import SubNav from './SubNav';
import SubNavFixture from './SubNav.fixture';

describe('<SubNav />', () => {
  it('should be defined', () => {
    expect(SubNav).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <SubNav heading={SubNavFixture.heading} links={SubNavFixture.links} />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without title if not `title` prop is passed', () => {
    const tree = renderer
      .create(<SubNav links={SubNavFixture.links} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
