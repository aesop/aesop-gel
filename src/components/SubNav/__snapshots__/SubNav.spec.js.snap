// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubNav /> renders base component correctly 1`] = `
<nav
  className="base dark"
>
  <h3
    className="base small dark heading"
  >
    Phasellus aliquet sem
  </h3>
  <ul
    className="base dark"
  >
    <li
      className="item"
    >
      <a
        className="base left dark item"
        href="https://aesop.com/"
        target="_self"
      >
        Morbi id nulla
      </a>
    </li>
    <li
      className="item"
    >
      <a
        className="base left dark item"
        href="https://aesop.com/"
        target="_self"
      >
        Nullam ultrices
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<SubNav /> renders without title if not \`title\` prop is passed 1`] = `
<nav
  className="base dark"
>
  <ul
    className="base dark"
  >
    <li
      className="item"
    >
      <a
        className="base left dark item"
        href="https://aesop.com/"
        target="_self"
      >
        Morbi id nulla
      </a>
    </li>
    <li
      className="item"
    >
      <a
        className="base left dark item"
        href="https://aesop.com/"
        target="_self"
      >
        Nullam ultrices
      </a>
    </li>
  </ul>
</nav>
`;
